Tutorial

magnus-madsen opened the pull request:
One pending issue is that `--tutorial` only works when executed from the JAR.

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,237 @@
+// that performs constant propagation for local variables. The program language
+// we consider is very simple, it has three types of statements:
+//
+//      x = c        [literal statement]
```
I was talking to Dan today, and he pointed out that "literal statement" isn't the name he would have used. A literal statement would be just `c`.

What we have (`x = c`) is more like an assignment statement, or literal assignment, if you want to be more specific, I guess.

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,237 @@
+    //
+    // We will represent the program-under-analysis using three input
+    // relations. One for each type of statement:
+    //
```
Might be useful to put end-of-line comments for these definitions, e.g.:

```
rel LitStm(r: Str, c: Int)      // r = c
```

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,237 @@
+
+    //
+    // This complete the specification of the lattice. To associate these
+    // functions with the type Constant we have to write some magic:
```
Let's not use the terms "magic" or "weird syntax."

We could just say "... we have to use the following syntax (which is subject to change in the future)"

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,237 @@
+    // an element of the constant propagation lattice. Strictly speaking,
+    // this is not necessary.
+    //
+    def alpha(i: Int): Constant = Constant.Cst(i)
```
Why is it called `alpha`? Also, when we say that it's not necessary, we should explain (maybe below) how it can be worked around.

magnus-madsen replied:
That is the abstract interpretation name. Alpha (abstraction) and gamma (concretization).

mhyee replied:
The comment should explain this. If you don't know abstract interpretation (clearly I don't), then `alpha` just comes from nowhere.

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,237 @@
+    def alpha(i: Int): Constant = Constant.Cst(i)
+
+    //
+    // We can now define the semantic rules of the analysis!
```
I was never a fan of sprinkling exclamation marks, but this is just a personal opinion.

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,237 @@
+    //
+
+    //
+    // The first rule models literal statements r = c and lifts the constant
```
I think a brief refresher on how "if the body is satisfied, then the head must be satisfied, i.e. we infer a new rule."

Or if we wrote out the rule in English. "If we have the literal statement `r = c`, then we infer that the LocalVar `r` has value `Cst(c)`."

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,237 @@
+        case _ => Constant.Top
+    }
+
+    // !! Notice the special case to avoid division-by-zero in the analysis itself!
```
Nitpick: I would probably format this as `!! Notice the special case ... analysis itself !!`. I think the symmetric exclamation marks makes it look nicer.

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,237 @@
+    //
+    // We encode this program using the input facts:
+    //
+    LitStm("x", 3).
```
Maybe repeat the input program as comments beside the input facts?

mhyee commented:
Overall the tutorial looks good. I made a bunch of comments, mostly minor.

A few more comments about the tutorial:

- I think it's worth having some facts that require a lub operation, i.e. use an input program with an if-statement and some value becoming top.

- I don't know if it's too much to put it all in one example, or if we should split up the example into multiple smaller ones.

- We might also want to explain the results in the tutorial, instead of just presenting them.

Now general comments:

- I assume the tutorial has been tested and everything works as expected?

- What happens with the `--tutorial` flag if it's not executed from the JAR? Graceful failure? Crash?

- I'm still not convinced about including the tutorial in the jar and then spitting it out. I can see the benefit -- it makes things easier for the user, who doesn't have to go find the URL and download the file. On the other hand, it strikes me as inelegant -- there's now another directory in `src` and we need another boilerplate object (`LocalResource`).

- We could just put the tutorial in the already-existing `examples` directory (since it is an example) and tell people to download it. And people who directly check out the source will get it anyway.

magnus-madsen commented:
I fixed your comments, although I did not do anything with lub and glb. I wonder how you can make that look reasonable in this small langauge.

I fixed it so --tutorial always works. We can talk about whether the embedding is a good idea.

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,233 @@
+    // We will represent the program-under-analysis using three input
+    // relations. One for each type of statement:
+    //
+    rel LitStm(r: Str, c: Int)         // r = c
```
If we renamed it to "assignment statement," then we should also change the name of the relation. It's still "LitStm" right now.

magnus-madsen replied:
Good point.

magnus-madsen commented:
The tutorial is now available here:

http://flix.github.io/doc/learn-flix.html

Better to have it there, than to have nothing.

mhyee commented on main/src/tutorial/tutorial.flix:
```diff
@@ -0,0 +1,233 @@
+    //
+    // The first rule models literal statements r = c and lifts the constant
+    // into the constant propagation lattice. Intuitively, if the body of
+    // the rule (whatever is to the right of ':-') then the head (what is on
```
Honestly, the `':-')` looks too much like an emoticon...

Maybe we should just annotate the rule directly?

```
LocalVar(r, alpha(c)) :- LitStm(r, c).
// ~~~~ head ~~~~~~~~    ~~~~ body ~~~~
```

mhyee commented:
Whoops, I hadn't finished my comments. In the future, if you ask me to review some code, let me say when I'm finished and think it's OK to merge.

I used the following example:

```
LitStm("a", 40).        // a = 40
LitStm("b", 1).         // b = 1
LitStm("c", 2).         // c = 2
                        // if (mystery()) {
AddStm("d", "a", "b").  //   d = a + b
                        // } else {
AddStm("d", "a", "c").  //   d = a + c
                        // } 
```

And so we get `LocalVar("d", Cst(41) lub Cst(42)) = LocalVar("d", Top)`.

More comments:

- When/if we get SBT to package jars, we should test that it works. But since the tutorial is in the src directory, I think it will.

- If we renamed it to "assignment statement," then we should also change the name of the relation. It's still "LitStm" right now.

magnus-madsen commented:
> Whoops, I hadn't finished my comments.

Yeah, sorry, I was rushed.

As for the example, the code sniplet is fine for the poster, but I wonder if it is intuitive enough in general. Would you not wonder why the if-statement is not represented?

mhyee commented:
That's a good point. I'm not sure I have an answer for that. But I still think it's important to have an example that uses lub.
